Sources/TensorFlow/Operators/Math.swift

```diff
@@ -2224,6 +2227,51 @@ internal extension Tensor where Scalar: TensorFlowFloatingPoint {
             ) / self
         })
     }
+
+    @inlinable
+    func _vjpProduct(squeezingAxes axes: Tensor<Int32>) -> (Tensor, (Tensor) -> Tensor) {
+        /// The gradient can be expressed by dividing the product by each entry of the
+        /// input tensor, but this approach can't deal with zeros in the input.
+        /// Here, we avoid this problem by composing the output as a product of two cumulativeProduct operations.
+        /// Adapted from `_ProdGrad` in Python TensorFlow:
+        /// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/math_grad.py
+        let result = product(squeezingAxes: axes)
+        return (result, { v in
+            // Reshape reduction indices for the case where the parameter is a scalar
+            var reductionIndices = axes.reshaped(to: TensorShape(-1))
+
+            // Expand grad to full input shape
+            let outputShapeKeptDims = self.shape.reducedShape(axes: reductionIndices.scalars)
```

Comment:
See other comment - please inline `reducedShape`.
```suggestion
            var outputShape = self.shape
            for axis in reductionIndices.scalars {
                outputShape[Int(axis)] = 1
            }
```